matlab/FeatherReader.cpp

```diff
@@ -0,0 +1,301 @@
+  std::shared_ptr<::arrow::io::RandomAccessFile> random_access_file(readable_file);
+  // Open the Feather file for reading with a TableReader.
+  HandleStatus(
+      ::arrow::ipc::feather::TableReader::Open(random_access_file, &table_reader_));
+}
+
+// Read the table metadata from the Feather file as a mxArray*.
+mxArray* FeatherReader::ReadMetadata() const {
+  const int num_metadata_fields = 4;
+  const char* fieldnames[] = {"NumRows", "NumVariables", "Description", "Version"};
+
+  // Create a mxArray struct array containing the table metadata to be passed back to
+  // MATLAB.
+  mxArray* metadata = mxCreateStructMatrix(1, 1, num_metadata_fields, fieldnames);
+
+  // Returning double values to MATLAB since that is the default type.
+  // Note: MATLAB's size on tables prevents importing a table any larger than 2^52 (many
+  // fewer rows/variables in fact)
+  // files with more data that will error earlier.
+
+  // Set the number of rows.
+  mxSetField(metadata, 0, "NumRows",
+             mxCreateDoubleScalar(static_cast<double>(num_rows_)));
+
+  // Set the number of variables.
+  mxSetField(metadata, 0, "NumVariables",
+             mxCreateDoubleScalar(static_cast<double>(num_variables_)));
+
+  // Set the description.
+  mxSetField(metadata, 0, "Description", mxCreateString(description_.c_str()));
+
+  // Set the version.
+  mxSetField(metadata, 0, "Version", mxCreateDoubleScalar(static_cast<double>(version_)));
+
+  return metadata;
+}
+
+// Read the table variables from the Feather file as a mxArray*.
+mxArray* FeatherReader::ReadVariables() const {
+  const int num_variable_fields = 4;
+  const char* fieldnames[] = {"Name", "Data", "Nulls", "Type"};
+
+  // Create an mxArray struct array containing the table variables to be passed back to
+  // MATLAB.
+  mxArray* variables =
+      mxCreateStructMatrix(1, num_variables_, num_variable_fields, fieldnames);
+
+  // Read all the table variables in the Feather file into memory.
+  for (int64_t i = 0; i < num_variables_; ++i) {
+    std::shared_ptr<::arrow::Column> column(nullptr);
+    HandleStatus(table_reader_->GetColumn(i, &column));
+
+    // set the struct fields data
+    mxSetField(variables, i, "Name", ReadVariableName(column));
+    mxSetField(variables, i, "Data", ReadVariableData(column));
+    mxSetField(variables, i, "Nulls", ReadVariableNulls(column));
+    mxSetField(variables, i, "Type", ReadVariableType(column));
+  }
+
+  return variables;
+}
+
+// Terminates execution of Feather file reading and returns to the MATLAB prompt,
+// displaying
+// an error message if the given status indicates that an error has occurred.
+void FeatherReader::HandleStatus(const ::arrow::Status& status) const {
+  const char* arrow_error_message = "Arrow error: %s";
+  switch (status.code()) {
+    case ::arrow::StatusCode::OK: {
+      break;
+    }
+    case ::arrow::StatusCode::OutOfMemory: {
+      mexErrMsgIdAndTxt("MATLAB:feather:read:arrow:status:OutOfMemory",
```

Comment:
Presumably we aspire to move beyond Feather files, in case you don't want to have "feather" hard-coded here